### ISSUE TITLE
ISFTPServer should define avatar = Attribute()

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6892]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#6892|
|Type|enhancement|
|Created|2014-01-07 11:32:20Z|
|Branch|https://github.com/twisted/twisted/tree/isftpserver-avatar-attr-6892|

The docstring clearly stated for ISFTPServer interface that

```
The only attribute of this class is "avatar". 
```

but the interface definition lacks the

```
class ISFTPServer(Interface):
  avatar = Attribute('The avatar returned by the Realm that we are authenticated with, and represents the logged-in user.')
```

This makes it harder to test implementations of ISFTPServer using zope.interface validation methods.

Attachments:

* [6892-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5d7/5d7b162d270753003f4f17dd2e0a8018eac9f0fc/8d641b62e57714881ed43a97ec683aec8fb13ec8.diff) (1972 bytes) - added by adiroiban on 2014-01-07 14:27:42Z - 
* [6892-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5d7/5d7b162d270753003f4f17dd2e0a8018eac9f0fc/fd36b530b337d566d810682d25e2ae4c29402ce3.diff) (1998 bytes) - added by adiroiban on 2014-01-22 11:07:19Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6892 6892
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__ 
branch__branches_isftpserver_avatar_attr_6892 branches/isftpserver-avatar-attr-6892
branch_author__julian julian
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1389094340000000 1389094340000000
changetime__1390438758000000 1390438758000000
version__None None
owner__Julian Julian
cc__alibotean cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban removed owner|
|-|-|

I have added avatar as an Attribute, wrote a test for the new code, updated docstring and created a news file.

Hope I did not forget anything.

I will attach the patch.

Real time diff is here https://github.com/chevah/twisted/compare/6892-ISFTPServer-avatar.diff


Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_3"></a>|@Julian set owner to @Julian|
|-|-|

Hey!

I'd add the word "communication" to the docstring summary, and also, misc topfiles should be empty. Both of those are minor though, I'll just apply that change and merge, think this looks good.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="julian's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|julian commented|
|-|-|

(In [41407]) Branching to isftpserver-avatar-attr-6892.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_5"></a>|@Julian set owner to @adiroiban|
|-|-|

Sorry, spoke too soon.

I applied the patch, and modulo some other fixes (misplaced paren, and a twistedchecker error), it seems like the test that was added [fails](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/isftpserver-avatar-attr-6892).

I'd suspect that probably what's meant is to check the `client` attribute on server, not server itself, but I'm going to volley this back just to be sure :).

Please resubmit with a fix, that one if it seems also to you to be the correct one after taking a look at the implementation there.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @Julian|
|-|-|

Sorry for the bad test. I have fixed it.

Not sure how I ended up with this. I think that somehow the error was shadowed by the known expectedFailures=2 for flatten tests.

I have attached a new diff.

Many thanks for the review.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_7"></a>|@Julian commented|
|-|-|

No worries, just wanted to make sure we were on the same page.

Thanks, looks good now!

**Comment by twisted-trac** [comment]
|<img alt="julian's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|julian set status to `closed`|
|-|-|

(In [41415]) Merge isftpserver-avatar-attr-6892: Add a missing attribute to the ISFTPServer interface.

Author: adiroiban
Reviewers: Julian
Fixes: [#6892](https://github.com/twisted/twisted/issues/6892)
